### PR TITLE
Supporting Workflow Projects

### PR DESCRIPTION
**Pull request by lukeknep**
The goal of this branch is to support GitLab Push Triggers in the Workflow project type.

The current state of this branch works for Workflow builds!

collaborating with @omehegan and @sudssm

### CODE REVIEW

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 1]
```diff
@@ -364,13 +384,13 @@ public File getLogFile() {
...
         public DescriptorImpl() {
-        	load();
+            load();
         }
         
         @Override
         public boolean isApplicable(Item item) {
-            if(item instanceof AbstractProject) {
-                project = (AbstractProject) item;
+            if(item instanceof Job) {
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: Probably also:

```java
&& item instanceof ParameterizedJob
```

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 2]
```diff
@@ -62,7 +66,7 @@
     // compatibility with earlier plugins
     public Object readResolve() {
         if (null == allowedBranches) {
-        	allowedBranches = new ArrayList<String>();
+            allowedBranches = new ArrayList<String>();
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: Gratuitous whitespace changes makes this hard to review.

Comment (lukeknep) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: @jglick Sorry for that, my editor wasn't set up correctly. I'll try to restore the whitespace.

**Comment by lukeknep** [comment]
@jglick I've cleaned up the whitespace. Thanks for taking a look!

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 3]
```diff
@@ -369,8 +388,8 @@ public DescriptorImpl() {
         
         @Override
         public boolean isApplicable(Item item) {
-            if(item instanceof AbstractProject) {
-                project = (AbstractProject) item;
+            if(item instanceof Job && item instanceof ParameterizedJobMixIn) {
```
Comment (lukeknep) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: @jglick Did you mean ParameterizedJobMixIn in your comment? I couldn't find ParameterizedJob in the Jenkins source.

Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: No. It is a nested type inside `ParameterizedJobMixIn`.

BTW [read this](https://github.com/jenkinsci/workflow-plugin/blob/master/COMPATIBILITY.md#historical-background).

**Comment by jaapio** [comment]
Any progress on this pr yet? We would love to have it integrated

**Comment by lukeknep** [comment]
@jaapio Yes! We have got the branch working for Workflow projects, and are just cleaning it up before asking for more comments.

**Comment by omehegan** [comment]
@jglick I think we need to bump the minimum required Jenkins version in the pom.xml to one where Workflow was introduced. Correct?

**Comment by jglick** [comment]
The plugin is already using 1.568, which might be enough, though I would advise just going up to 1.580 or 1.580.1.

Needs a merge with `master` to resolve conflicts.

**Comment by lukeknep** [comment]
The merge from master unfortunately broke the Workflow functionality. Currently my version is unable to find any SCMs on the Workflow build. This is the configuration I'm using:

```
node () {
checkout([$class: 'GitSCM', branches: [[name: '**']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[url: 'git@gitlab.btrll.com:productivity/build_tools.git']]])
}
```

**Comment by lukeknep** [comment]
@jglick The plugin is working great for us, for both Workflow and Freestyle projects, and I think it's ready for review. Would you have time to give it another look?

**Comment by lukeknep** [comment]
Need to squash commits.

**Comment by omehegan** [comment]
@bassrock do you have any comment on this?

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 4]
```diff
@@ -158,8 +161,16 @@ public String getExcludeBranchesSpec() {
...
 
+    // executes when the Trigger receives a push request
     public void onPost(final GitLabPushRequest req) {
-    	if (triggerOnPush && (allowAllBranches || this.isBranchAllowed(this.getSourceBranch(req)))) {
+    	final ParameterizedJobMixIn scheduledJob = new ParameterizedJobMixIn() {
+            @Override
+            protected Job asJob() {
+                return job;
+            }
+        };
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: BTW as of 1.621 there is a `static` convenience method to do this mess. My practice is to add a comment

```java
// TODO 1.621+ use standard method
```

as a reminder for the next person to update the core baseline.

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 5]
```diff
@@ -168,7 +179,18 @@ public void run() {
             		String name = " #" + job.getNextBuildNumber();
             		GitLabPushCause cause = createGitLabPushCause(req);
             		Action[] actions = createActions(req);
-            		if (job.scheduleBuild(job.getQuietPeriod(), cause, actions)) {
+
+                    boolean scheduled;
+
+                    if (job instanceof AbstractProject<?,?>) {
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: No need for two code paths. `ParameterizedJobMixIn.scheduleBuild2` should have everything you need, including `Action` list support.

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 6]
```diff
@@ -203,7 +225,11 @@ private GitLabPushCause createGitLabPushCause(GitLabPushRequest req) {
                     values.put("gitlabTargetBranch", new StringParameterValue("gitlabTargetBranch", branch));
                     values.put("gitlabBranch", new StringParameterValue("gitlabBranch", branch));
 
-                    LOGGER.log(Level.INFO, "Trying to get name and URL for job: {0} using project {1} (push)", new String[]{job.getName(), job.getRootProject().getName()});
+                    if (job instanceof AbstractProject<?,?>){
+                        LOGGER.log(Level.INFO, "Trying to get name and URL for job: {0} using project {1} (push)", new String[]{job.getName(), ((AbstractProject<?, ?>) job).getRootProject().getName()});
+                    }else{
+                        LOGGER.log(Level.INFO, "Trying to get name and URL for job: {0} (push)", new String[]{job.getName()});
+                    }
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: BTW (not related to this PR): `getFullName` gives a complete path when using the folders plugin.

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 7]
```diff
@@ -242,7 +269,23 @@ public void run() {
...
+                    };
+
+                    boolean scheduled;
+                    if (job instanceof AbstractProject<?,?>) {
+                        AbstractProject job_ap = (AbstractProject<?, ?>) job;
+                        scheduled = job_ap.scheduleBuild(job_ap.getQuietPeriod(), cause, actions);
+                    }
+                    else {
+                        scheduled = scheduledJob.scheduleBuild(cause);
+                    }
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: ditto here

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabRunListener.java [hunk 8]
```diff
@@ -14,31 +15,33 @@
...
-    private GitLabPushTrigger getTrigger(AbstractBuild abstractBuild) {
-        Trigger trig = abstractBuild.getProject().getTrigger(GitLabPushTrigger.class);
-        if (trig != null && trig instanceof GitLabPushTrigger)
-            return (GitLabPushTrigger) trig;
+    private GitLabPushTrigger getTrigger(Run run) {
+        if (run instanceof AbstractBuild) {
+            Trigger trig = ((AbstractBuild) run).getProject().getTrigger(GitLabPushTrigger.class);
+            if (trig != null && trig instanceof GitLabPushTrigger)
+                return (GitLabPushTrigger) trig;
+        }
```
Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabRunListener.java]: Again see the convenience method in `ParameterizedJobMixIn` in 1.621 for getting a specific trigger from a job (including Workflow), and in the meantime inline that idiom here.

Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabRunListener.java]: (You actually got it right in `GitLabWebHook`, maybe forgot to use that here?)

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java (hunk 5 above)
Comment (sudssm) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: @jglick 
scheduleBuild2 doesn't take a GitLabPushCause parameter?
Also, in the case of the job not being an AbstractProject (so getQuietPeriod isn't available), would 0 be the best choice for the quietPeriod?

Comment (jglick) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java]: > scheduleBuild2 doesn't take a GitLabPushCause parameter?

Just wrap with `new CauseAction(…)`.



> in the case of the job not being an AbstractProject (so getQuietPeriod isn't available)

See the convenience method in 1.621: use `ParameterizedJob.getQuietPeriod`.

**Comment by omehegan** [comment]
@lukeknep can you close this since we got this merged in the main repo?

**Comment by omehegan** [comment]
@lukeknep got this merged upstream, closing this PR.

**Pull request closed without merging**